packages/runtime/container-runtime/src/summarizerClientElection.ts

```diff
@@ -6,88 +6,120 @@
-    private lastSummaryAckSeqForClient = 0;
-    private _hasAnySummarizersInQuorum: boolean;
-    private hasLoggedTelemetry = false;
+ export class SummarizerClientElection extends TypedEventEmitter<ISummarizerClientElectionEvents> {
+    /**
+     * Used to calculate number of ops since last summary ack for the current elected client.
+     * This will be undefined if there is no elected summarizer, or no summary ack has been
+     * observed since this client was elected.
+     * When a summary ack comes in, this will be set to the sequence number of the summary ack.
+     */
+    private lastSummaryAckSeqForClient: number | undefined;
 
     public get electedClientId() {
-        return this.clientElection.getElectedClient()?.clientId;
-    }
-
-    public get hasSummarizersInQuorum() {
-        return this._hasAnySummarizersInQuorum;
+        return this.clientElection.electedClient?.clientId;
     }
 
     constructor(
         private readonly logger: ITelemetryLogger,
         private readonly summaryCollection: IEventProvider<ISummaryCollectionOpEvents>,
-        public readonly clientElection: OrderedClientElection,
+        public readonly clientElection: IOrderedClientElection,
         private readonly maxOpsSinceLastSummary: number,
     ) {
         super();
-        this.summaryCollection.on("default", (op) => {
-            const opsSinceLastAckForClient = op.sequenceNumber - this.lastSummaryAckSeqForClient;
-            if (
-                opsSinceLastAckForClient > this.maxOpsSinceLastSummary
-                && !this.hasLoggedTelemetry
-                && this.electedClientId !== undefined
-            ) {
-                // Limit telemetry to only next client?
+        // On every inbound op, if enough ops pass without seeing a summary ack (per elected client),
+        // elect a new client and log to telemetry.
+        this.summaryCollection.on("default", ({ sequenceNumber }) => {
+            const electedClientId = this.electedClientId;
+            if (electedClientId === undefined) {
+                // Reset election if no elected client, but eligible clients are connected.
+                // This should be uncommon, but is possible if the initial state of the
+                // ordered client election contains an undefined client id or one not found
+                // in the quorum (the latter would already log an error).
+                if (this.clientElection.eligibleCount > 0) {
+                    this.clientElection.resetElectedClient(sequenceNumber);
+                }
+                return;
+            }
+            let electionSequenceNumber = this.clientElection.electionSequenceNumber;
+            const opsWithoutSummary = sequenceNumber - (this.lastSummaryAckSeqForClient ?? electionSequenceNumber);
+            if (opsWithoutSummary > this.maxOpsSinceLastSummary) {
+                // Log and elect a new summarizer client.
                 this.logger.sendErrorEvent({
```

Comment:
i'm worried this could emit alot of logs as once we pass the threshold it will emit for every op until new election or summary succeeds

Reply:
This is prevented by all the code being synchronous. If you look below, I added an event to detect where this scenario could appear. It might be better to add an extra layer of protection against this though...

The intended flow on receiving an op is this:
1. Use op sequence number (`N`) to calculate `opsWithoutSummary`
2. If `opsWithoutSummary` > threshold, then send ElectedClientNotSummarizing event
3. Also increment elected client- this is synchronous
    1. Should set `electionSequenceNumber` to the op sequence number (`N`)
    2. Should emit an "election" event
4. We handle the "election" event synchronously
    1. Should unset `lastSummaryAckSeqForClient` to undefined
    2. If it was the last client in the quorum, we reset but still update `electionSequenceNumber`
5. Now resuming our op handler, we essentially recalculate `opsWithoutSummary` like so:
    1. `= op sequence number - (lastSummaryAckSeqForClient ?? electionSequenceNumber)`
    2. `= N - (undefined ?? N)`
    3. `= 0` Otherwise we throw an "UnexpectedElectionSequenceNumber" error.

I think this intended flow is okay, but I will add an extra safeguard such that in case it does not work as intended, instead of just logging an error message, we will also prevent excessive per-op telemetry logs.

Reply:
I made a change to help prevent excess logging. Let me know if it seems okay.